File: src/test/java/ru/anseranser/peshki/ai/BotStrategyTest.java
package ru.anseranser.peshki.ai;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.anseranser.peshki.engine.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class BotStrategyTest {

    private GameConfig config;
    private Board board;
    private BotStrategy strategy;

    @BeforeEach
    void setUp() {
        config = GameConfig.DEFAULT;
        board = new Board(config);
        strategy = new BotStrategy();
    }

    @Test
    void selectBestMoveReturnsNonNull() {
        Player player = board.getPlayers().get(1);
        // Place a pawn on the field
        Pawn pawn = player.getPawns().getFirst();
        Cell fieldCell = board.getCorner(player).getNextFieldCell();
        pawn.setCell(fieldCell);
        pawn.setState(Pawn.State.FIELDER);
        fieldCell.setPawn(pawn);

        List<Move> moves = GameEngine.generateAllMoves(player, List.of(3, 4), board, config);
        if (!moves.isEmpty()) {
            Move best = strategy.selectBestMove(player, moves, config);
            assertNotNull(best);
        }
    }

    @Test
    void selectBestMoveReturnsHighestScored() {
        Player player = board.getPlayers().get(1);
        Player enemy = board.getPlayers().get(0);

        // Place my pawn on field
        Pawn myPawn = player.getPawns().getFirst();
        Cell myCell = board.getCorner(player).getNextFieldCell();
        myPawn.setCell(myCell);
        myPawn.setState(Pawn.State.FIELDER);
        myCell.setPawn(myPawn);

        // Place enemy pawn 1 step ahead
        Pawn enemyPawn = enemy.getPawns().getFirst();
        Cell enemyCell = myCell.getNextFieldCell();
        enemyPawn.setCell(enemyCell);
        enemyPawn.setState(Pawn.State.FIELDER);
        enemyCell.setPawn(enemyPawn);

        // Create moves: one that kills, one that doesn't
        Move killMove = new Move(myPawn, 1, List.of(1));
        Move normalMove = new Move(player.getPawns().get(1), 1, List.of(1));

        // Place second pawn for normal move
        Pawn pawn2 = player.getPawns().get(1);
        Cell cell2 = board.getCorner(player).getNextFieldCell().getNextFieldCell().getNextFieldCell();
        pawn2.setCell(cell2);
        pawn2.setState(Pawn.State.FIELDER);
        cell2.setPawn(pawn2);

        Move best = strategy.selectBestMove(player, List.of(killMove, normalMove), config);
        assertNotNull(best);
        assertEquals(killMove, best);
    }

    @Test
    void botDoesNotCrashWithEmptyMoves() {
        Player player = board.getPlayers().get(1);
        Move best = strategy.selectBestMove(player, List.of(), config);
        assertNull(best);
    }
}

File: web/tests/BotStrategy.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { DEFAULT_CONFIG } from '../src/engine/GameConfig';
import { Board } from '../src/engine/Board';
import { PawnState } from '../src/engine/Pawn';
import { GameEngine } from '../src/engine/GameEngine';
import { BotStrategy } from '../src/ai/BotStrategy';

describe('BotStrategy', () => {
  let board: Board;
  let strategy: BotStrategy;

  beforeEach(() => {
    board = new Board(DEFAULT_CONFIG);
    strategy = new BotStrategy();
  });

  it('selects best move from available', () => {
    const player = board.players[1];
    player.pawns[0].cell = board.getCorner(player).nextFieldCell;
    player.pawns[0].state = PawnState.FIELDER;
    board.getCorner(player).nextFieldCell!.pawn = player.pawns[0];

    const moves = GameEngine.generateAllMoves(player, [3, 4], board, DEFAULT_CONFIG);
    const best = strategy.selectBestMove(player, moves, DEFAULT_CONFIG);
    expect(best).not.toBeNull();
  });

  it('prefers killing move', () => {
    const player = board.players[0];
    const enemy = board.players[1];

    const myPawn = player.pawns[0];
    const cell = board.getCorner(player).nextFieldCell!;
    myPawn.cell = cell;
    myPawn.state = PawnState.FIELDER;
    cell.pawn = myPawn;

    const enemyPawn = enemy.pawns[0];
    const targetCell = cell.nextFieldCell!;
    enemyPawn.cell = targetCell;
    enemyPawn.state = PawnState.FIELDER;
    targetCell.pawn = enemyPawn;

    const moves = GameEngine.generateAllMoves(player, [1, 6], board, DEFAULT_CONFIG);
    const best = strategy.selectBestMove(player, moves, DEFAULT_CONFIG);
    expect(best).not.toBeNull();
    if (best && best.pawn) {
      expect(best.pawn).toBe(myPawn);
      expect(best.steps).toBe(1);
    }
  });

  it('returns null for empty moves', () => {
    const player = board.players[0];
    const best = strategy.selectBestMove(player, [], DEFAULT_CONFIG);
    expect(best).toBeNull();
  });
});
